A29-xds-tls-security.md

```diff
@@ -0,0 +1,422 @@
+
+```C++
+   // TODO: yashykt to add
+```
+
+#### Java
+
+Use of XdsChannelCredentials:
+
+```Java
+    ChannelCredentials credentials = XdsChannelCredentials.create(InsecureChannelCredentials.create());
+    ManagedChannel channel = Grpc.newChannelBuilder(target, credentials).build();
+```
+
+Use of XdsServerCredentials:
+
+```Java
+    ServerCredentials credentials = XdsServerCredentials.create(InsecureServerCredentials.create());
+    Server server = XdsServerBuilder.forPort(port, credentials)
+      .addService(new HostnameGreeter(hostname))
+      .build()
+      .start();
+```
+
+#### Go
+
+Use of XdsChannelCredentials:
+
+```Go
+   // TODO: easwars to add
+```
+
+Use of XdsServerCredentials:
+
+```Go
+   // TODO: easwars to add
+```
+
+### xDS Protocol
+
+gRPC uses CDS to configure client side security and LDS for server side security as described below.
+The required fields were added to xDS v3 hence xDS v3 is a pre-requisite for this proposal.
+
+#### CDS for Client Side Security
+
+The CDS policy is the top level LB policy applied to all the connections under that policy.
+[A27:CDS][] describes the flow. The field [`transport_socket`][CL-TS] is used to extract the
+[`UpstreamTlsContext`][UTC] as described [here][CL-TS-comment].
+Note that we don't (currently) support [`transport_socket_matches`][CL-TS-matches].
+The `UpstreamTlsContext` thus obtained is passed down to all the child policies and connections.
+How this is done is language dependent and is similar to how the implementations pass policy
+information down to child policies.
+
+[`common_tls_context`][CTC] in the `UpstreamTlsContext` contains the required security configuration.
+All other fields are ignored. See below for [`CommonTlsContext`][CTC-type] processing.
+
+[A27:CDS]: A27-xds-global-load-balancing.md#cds
+[CL-TS]: https://github.com/envoyproxy/envoy/blob/9aca65c395f01020080166e4795455addde167fa/api/envoy/config/cluster/v3/cluster.proto#L937
+[CL-TS-comment]: https://github.com/envoyproxy/envoy/blob/9aca65c395f01020080166e4795455addde167fa/api/envoy/config/cluster/v3/cluster.proto#L932
+[CL-TS-matches]: https://github.com/envoyproxy/envoy/blob/9aca65c395f01020080166e4795455addde167fa/api/envoy/config/cluster/v3/cluster.proto#L680
+[CTC]: https://github.com/envoyproxy/envoy/blob/7d4b2cae486b66b62ba0d3e1e348504699bea1bf/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L37
+
+#### LDS for Server Side Security
+
+Server-side xDS processing is described in [A36: xDS-Enabled Servers][A36].
+An LDS may have one or more [`filter_chains`][filter-chains] and one of those
+`filter_chain`s is matched against an incoming connection in order to get and apply
+the security (and other) configuration for that connection as described in the
+[`FilterChainMatch`][filter-chain-match]. The [`transport_socket`][transport-socket]
+of the matched (selected) `filter_chain` is used to extract the
+[`DownstreamTlsContext`][DTC] as described [here][transport-socket-comment].
```

Comment:
gRPC's behavior on decoding `transport_socket` is slightly different where if the name is not `envoy.transport_sockets.tls`, the overall behavior is to use the fallback creds instead of plaintext

Reply:
Good point. Ideally we should fix the xDS proto file to include the gRPC variation (and make all of them less Envoy centric in general).

Reply:
> ...Why do you use fallback creds in this case?

There are arguments for both sides although we should have consistent behavior in all languages. One could argue that we don't recognize anything other than `envoy.transport_sockets.tls` so we NACK the update (like Go is doing). On the other hand we can define for gRPC the behavior that the fallback creds are used when we don't understand the `transport_socket`. My vote is for using the fallback creds but will agree to NACK if there are more convincing arguments.

Reply:
Using fallback creds when the name field is anything other than `envoy.transport_sockets.tls` seems like a special-case behavior, while NACK'ing is consistent with what we do everywhere else (wherein, if we receive something that we don't like or support, we NACK the resource).

Reply:
Okay NACK it is then. I'll update accordingly.